### PR TITLE
Maintain backward and forward compatibility during state and version file migrations. Closes #1913

### LINKED ISSUE #1913: Maintain backward and forward compatibility during state and version file migrations
```
// State is a struct containing installation state
type State struct {
	LegacyLastCheck      string `json:"lastChecked"`     // an RFC3339 encoded time stamp
	LegacyInstallationID string `json:"installationId"`  // a UUIDv4 string
	LastCheck            string `json:"last_checked"`    // an RFC3339 encoded time stamp
	InstallationID       string `json:"installation_id"` // a UUIDv4 string
	StructVersion        int64  `json:"struct_version"`
}
```
```
{
 "lastChecked": "2022-04-21T13:34:24+05:30",
 "installationId": "d68e1d96-2cb7-4592-ac40-1a31856e5549",
 "last_checked": "2022-04-21T13:39:01+05:30",
 "installation_id": "d68e1d96-2cb7-4592-ac40-1a31856e5549",
 "struct_version": 20220411
}
```

### PR DESCRIPTION
**Pull request by pskrbasu**

### CODE REVIEW

File: cmd/root.go [hunk 1]
```diff
@@ -150,12 +147,10 @@ func migrateLegacyFiles() error {
 		return nil
 	}
 	return utils.CombineErrors(
-		migrate.Migrate(statefile.LegacyState{}, &statefile.State{}, filepaths.LegacyStateFilePath()),
-		migrate.Migrate(pluginmanager.LegacyPluginManagerState{}, &pluginmanager.PluginManagerState{}, filepaths.PluginManagerStateFilePath()),
-		migrate.Migrate(db_local.LegacyRunningDBInstanceInfo{}, &db_local.RunningDBInstanceInfo{}, filepaths.RunningInfoFilePath()),
-		migrate.Migrate(dashboardserver.LegacyDashboardServiceState{}, &dashboardserver.DashboardServiceState{}, filepaths.DashboardServiceStateFilePath()),
-		migrate.Migrate(versionfile.LegacyPluginVersionFile{}, &versionfile.PluginVersionFile{}, filepaths.PluginVersionFilePath()),
-		migrate.Migrate(versionfile.LegacyDatabaseVersionFile{}, &versionfile.DatabaseVersionFile{}, filepaths.DatabaseVersionFilePath()),
+		migrate.Migrate(&statefile.State{}, filepaths.LegacyStateFilePath()),
```
Comment (kaidaguerre) [inline cmd/root.go:150]: do we need to migrate state file? I thought not?

File: migrate/migrate.go [hunk 2]
```diff
@@ -8,36 +8,30 @@ import (
...
 
 type Migrateable interface {
-	MigrateFrom(old interface{}) Migrateable
+	MigrateFrom() Migrateable
 	IsValid() bool
 	Save() error
 }
 
-func Migrate[O any, T Migrateable](old O, new T, oldPath string) error {
+func Migrate(old Migrateable, oldPath string) error {
```
Comment (kaidaguerre) [inline migrate/migrate.go:16]: is `old` the best variable name now?
maybe just `migrateable` ?

File: ociinstaller/versionfile/installed_version.go [hunk 3]
```diff
@@ -1,22 +1,28 @@
...
-	InstallDate        string `json:"install_date"`
+	ImageDigest        string `json:"image_digest,omitempty"`
+	BinaryDigest       string `json:"binary_digest,omitempty"`
+	BinaryArchitecture string `json:"binary_arch,omitempty"`
+	InstalledFrom      string `json:"installed_from,omitempty"`
+	LastCheckedDate    string `json:"last_checked_date,omitempty"`
+	InstallDate        string `json:"install_date,omitempty"`
+
+	// json tags are same in the new version
+	// LegacyName            string `json:"name"`
```
Comment (kaidaguerre) [inline ociinstaller/versionfile/installed_version.go:14]: remove commented out code, not needed

File: steampipeconfig/modconfig/connection.go [hunk 4]
```diff
@@ -67,6 +42,29 @@ type Connection struct {
...
 	Options   *options.Connection `json:"options,omitempty"`
 	DeclRange Range               `json:"decl_range,omitempty"`
+
+	// legacy properties
+	LegacyName            string                 `json:"Name,omitempty"`
+	LegacyPluginShortName string                 `json:"PluginShortName,omitempty"`
+	LegacyPlugin          string                 `json:"Plugin,omitempty"`
+	LegacyType            string                 `json:"Type,omitempty"`
+	LegacyConnectionNames []string               `json:"Connections,omitempty"`
+	LegacyConnections     map[string]*Connection `json:"-"`
```
Comment (kaidaguerre) [inline steampipeconfig/modconfig/connection.go:52]: why have a legacy property which is not exported to JSON?

File: steampipeconfig/versionmap/workspace_lock.go [hunk 5]
```diff
@@ -48,10 +48,9 @@ func (s WorkspaceLock) IsValid() bool {
 	return s.StructVersion > 0
 }
 
-func (s *WorkspaceLock) MigrateFrom(prev interface{}) migrate.Migrateable {
-	legacyState := prev.(DependencyVersionMap)
+func (s *WorkspaceLock) MigrateFrom() migrate.Migrateable {
 	s.StructVersion = WorkspaceLockStructVersion
-	s.InstallCache = legacyState
+	// s.InstallCache = legacyState
```
Comment (kaidaguerre) [inline steampipeconfig/versionmap/workspace_lock.go:53]: remove

File: workspace/workspace.go [hunk 6]
```diff
@@ -76,8 +75,8 @@ func Load(ctx context.Context, workspacePath string) (*Workspace, error) {
 	}
 
 	// migrate legacy workspace lock files in the directory to use snake casing (migrated in v0.14.0)
-	err := migrate.Migrate(versionmap.DependencyVersionMap{}, &versionmap.WorkspaceLock{}, filepaths.WorkspaceLockPath(workspacePath))
-	utils.FailOnErrorWithMessage(err, "failed to migrate legacy workspace lock files")
+	// err := migrate.Migrate(&versionmap.WorkspaceLock{}, filepaths.WorkspaceLockPath(workspacePath))
```
Comment (kaidaguerre) [inline workspace/workspace.go:78]: to be uncommented?

**Review by kaidaguerre — CHANGES_REQUESTED** [review]

File: cmd/root.go (hunk 1 above)
Comment (pskrbasu) [reply cmd/root.go:150]: this is the update_check.json file

File: statefile/state.go [hunk 7]
```diff
@@ -15,18 +15,13 @@ import (
...
-}
-
 // State is a struct containing installation state
 type State struct {
-	LastCheck      string `json:"last_checked"`    // an RFC3339 encoded time stamp
-	InstallationID string `json:"installation_id"` // a UUIDv4 string
-	StructVersion  int64  `json:"struct_version"`
+	LastCheck            string `json:"last_checked"`    // an RFC3339 encoded time stamp
+	InstallationID       string `json:"installation_id"` // a UUIDv4 string
+	StructVersion        int64  `json:"struct_version"`
```
Comment (kaidaguerre) [inline statefile/state.go:22]: Update all legacy property declarations to be like this: 
```
type State struct {
	LastCheck            string `json:"last_checked"`    // an RFC3339 encoded time stamp
	InstallationID       string `json:"installation_id"` // a UUIDv4 string
	StructVersion        int64  `json:"struct_version"`

        // legacy properties included for backwards compatibility with v0.13
	LegacyLastCheck      string `json:"lastChecked"`   
	LegacyInstallationID string `json:"installationId"` 
}
```

File: steampipeconfig/modconfig/connection.go (hunk 4 above)
Comment (kaidaguerre) [reply steampipeconfig/modconfig/connection.go:52]: ???? ^^^^^

**Review by kaidaguerre — CHANGES_REQUESTED** [review]

**Pull request merged**